File: android2_lesson01/src/java/java/com/example/ama/android2_lesson01/widget/NotesWidgetProvider.java
package com.example.ama.android2_lesson01.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;

import com.example.ama.android2_lesson01.R;
import com.example.ama.android2_lesson01.ui.MainActivity;

public class NotesWidgetProvider extends AppWidgetProvider {

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        updateWidgets(context);
        super.onUpdate(context, appWidgetManager, appWidgetIds);
    }

    public static void updateWidgets(Context context) {
        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
        int[] IDs = appWidgetManager.getAppWidgetIds(new ComponentName(context, NotesWidgetProvider.class));
        for (int ID : IDs) {
            Intent intent = new Intent(context, ListNoteService.class);
            RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.widget_notes);
            remoteViews.setRemoteAdapter(R.id.lv_notelist, intent);

            Intent activityIntent = new Intent(context, MainActivity.class);
            PendingIntent pendingIntent = PendingIntent
                    .getActivity(context, 0, activityIntent, PendingIntent.FLAG_UPDATE_CURRENT);
            remoteViews.setPendingIntentTemplate(R.id.lv_notelist, pendingIntent);

            appWidgetManager.notifyAppWidgetViewDataChanged(ID, R.id.lv_notelist);
            appWidgetManager.updateAppWidget(ID, remoteViews);
        }
    }
}

File: android2_lesson01/src/kotlin/java/com/example/ama/android2_lesson01/widget/NotesWidgetProvider.kt
package com.example.ama.android2_lesson01.widget

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import com.example.ama.android2_lesson01.R

class NotesWidgetProvider : AppWidgetProvider() {

    companion object {
        fun updateWidget(context: Context) {
            val appWidgetManager = AppWidgetManager.getInstance(context)
            val widgetIds = appWidgetManager
                    .getAppWidgetIds(ComponentName(context, NotesWidgetProvider::class.java))
            for (id in widgetIds) {
                val remoteViews = RemoteViews(context.packageName, R.layout.widget_notes)
                remoteViews.setRemoteAdapter(R.id.lv_notelist, Intent(context, ListNoteService::class.java))
                appWidgetManager.notifyAppWidgetViewDataChanged(id, R.id.lv_notelist)
                appWidgetManager.updateAppWidget(id, remoteViews)
            }
        }
    }

    override fun onUpdate(context: Context?, appWidgetManager: AppWidgetManager?, appWidgetIds: IntArray?) {
        if (context != null) {
            updateWidget(context)
        }
        super.onUpdate(context, appWidgetManager, appWidgetIds)
    }
}